### PR TITLE
Deploy: Deloyment of private charms

### PR DESCRIPTION
**Pull request by SimonRichardson**
When resolving a private charm we need to correctly discharge a
macaroon, so that we can get a token. When we moved the resolving of
charms to the server-side, this was taken into account. In the end, we
believed that it would be better to fall back to the previous method of
charm resolution. This is until server-side deployments will be a thing.

The code path for this was wrong in the client. The new strategy for 
charm resolution now states; if there is no error, then just return. Else if it's
a charmhub charm and it's not supported show a nice error. Otherwise,
just fail. If it's any other charm, fallback to the prior workflow. 

As we have no way to know if a charm is a private charm upfront without
an API call (causing a discharge), the scenario is the best we can achieve
when dealing with multiple stores.

### Setup private charm

Note: you'll have to fix the dummy charm by adding some series to the
metadata.yaml

```sh
$ charm login
$ charm push ./testcharms/charm-repo/quantal/dummy
$ charm release cs:~<user-name>/dummy-1
```

Note: don't grant permissions!

### Deploy private charm

```sh
$ juju bootstrap lxd test
$ juju deploy cs:~<user-name>/dummy-1
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1932072

### CODE REVIEW

File: cmd/juju/application/store/charmadapter_test.go [hunk 1]
```diff
@@ -109,8 +110,8 @@ func (s *resolveSuite) TestResolveCharmFailResolveWithChannel(c *gc.C) {
 		return s.downloadClient, nil
 	})
 	_, obtainedOrigin, _, err := charmAdapter.ResolveCharm(curl, origin, false)
-	c.Assert(err, gc.NotNil)
-	c.Assert(obtainedOrigin.Risk, gc.Equals, string(csparams.NoChannel))
+	c.Assert(err, gc.IsNil)
```
Comment (manadart) [inline cmd/juju/application/store/charmadapter_test.go:113]: Do we then need to change the test name from `TestResolveCharmFailResolveWithChannel`? It's not a fail scenario now.

**Review by manadart — APPROVED** [review]
Works. One nit.

File: cmd/juju/application/store/charmadapter_test.go (hunk 1 above)
Comment (SimonRichardson) [reply cmd/juju/application/store/charmadapter_test.go:113]: So it is a ResolveCharm fail, but a success fallback scenario. I'll amend.

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**